Structure_Engine/Query/MaterialComposition.cs

```diff
@@ -308,6 +308,38 @@ public static MaterialComposition MaterialComposition(this BuiltUpRibbed propert
             new List<double> { property.TopThickness, volPerAreaRibZone });
         }
 
+        /***************************************************/
+
+        [Description("Returns a SurfaceProperty's MaterialComposition.")]
+        [Input("property", "The SurfaceProperty to query.")]
+        [Input("reinforcementDensity", "ReinforcementDensity assigned to the panel.")]
+        [Output("materialComposition", "The MaterialComposition of the SurfaceProperty.")]
+        public static MaterialComposition MaterialComposition(this BuiltUpDoubleRibbed property, ReinforcementDensity reinforcementDensity = null)
+        {
+            if (property.IsNull() || property.Material.IsNull())
+                return null;
+
+            //If only main material provided, use it for all parts
+            if (property.RibMaterial == null)
+                return property.Material.MaterialComposition(reinforcementDensity);
+
+            IMaterialFragment topMat = property.Material;
+            IMaterialFragment ribMat = property.RibMaterial ?? property.Material;
+
+            bool ribIsConcrete = ribMat is Concrete;
+            bool topIsConcrete = topMat is Concrete;
+
+            bool reinfToRib = ribIsConcrete || ribIsConcrete == topIsConcrete;  //Add reinforcement to ribs if they are of concrete material type or if non is concrete
+            bool reinfToSlab = topIsConcrete || ribIsConcrete == topIsConcrete; //Add reinforcement to slabs if it is of concrete material type or if non is concrete
```

Comment:
I am little confused with this bit, it looks like we are accounting for `ReinforcementDensity` in the slab and/or ribs depending on whether the `MaterialFragment` is `Concrete`.

However, in your test script if both rib and slab is `Timber`, then the composition of the reinforcement is the same as if it were all `Concrete`. Presumably it should be zero?

Reply:
As discussed offline, will add comments outlining my logic here. Can also record a note to the user how the reinforcement density has been applied for the cases where one of the materials is not concrete.

Reply:
Have refactored this part, leaving functionality as is, but with more comments and messaging to the user regarding what is happening.